framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,601 @@
+            // composed of digits.)
+            if (basename.contains(".")) {
+                String innermostClassname = basename.substring(basename.lastIndexOf('.') + 1);
+                if (innermostClassname.matches("\\d+")) {
+                    continue;
+                }
+            }
+
+            if (!"".equals(pkg)) {
+                printWriter.println("package " + pkg + ";");
+            }
+            if ("package-info".equals(basename) || "module-info".equals(basename)) {
+                continue;
+            }
+
+            // At this point, we no longer care about the distinction between packages
+            // and inner classes, so we should replace the $ in the definition of any
+            // inner classes with a ., so that they are printed correctly in stub files.
+
+            String fullyQualifiedClassname = Signatures.binaryNameToFullyQualified(classname);
+
+            int curlyCount =
+                    1
+                            + printClassDefinitions(
+                                    basename, fullyQualifiedClassname, aClass, printWriter);
+
+            // print fields or enum constants
+            if (!enumConstants.containsKey(fullyQualifiedClassname)) {
+                printFields(aClass, fullyQualifiedClassname, printWriter);
+            } else {
+                // for enums, instead of printing fields print the enum constants
+                List<VariableElement> enumConstants =
+                        this.enumConstants.get(fullyQualifiedClassname);
+
+                StringJoiner sj = new StringJoiner(", ");
+                for (VariableElement enumConstant : enumConstants) {
+                    sj.add(enumConstant.getSimpleName());
+                }
+                if (sj.length() != 0) {
+                    printWriter.print(sj.toString());
+                    printWriter.println(";");
+                }
+            }
+
+            // print method signatures
+            for (Map.Entry<String, AMethod> methodEntry : aClass.methods.entrySet()) {
+                AMethod aMethod = methodEntry.getValue();
+                printMethodSignature(aMethod, basename, fullyQualifiedClassname, printWriter);
+            }
+            for (int i = 0; i < curlyCount; i++) {
+                printWriter.println("}");
+            }
+        }
+        printWriter.flush();
+    }
+
+    /**
+     * Prints the type parameters of the given class, enclosed in {@code <...>}.
+     *
+     * @param classname a fully-qualified class name
+     * @param printWriter where to print the type parameters
+     */
+    private void printTypeParameters(String classname, PrintWriter printWriter) {
+        TypeElement type = types.get(classname);
+        if (type == null) {
+            return;
+        }
+        List<? extends TypeParameterElement> typeParameters = type.getTypeParameters();
+        if (typeParameters == null || typeParameters.isEmpty()) {
```

Comment:
`getTypeParameters` is specified to return an empty list rather than null.  Please don't do needless tests.